src/components/component.ts

```diff
@@ -479,29 +480,51 @@ module Plottable {
      * @returns The calling Component.
      */
     public detach() {
+      this.parent(null);
+
       if (this._isAnchored) {
         this._element.remove();
       }
+      this._isAnchored = false;
 
-      var parent: ComponentContainer = this._parent();
+      this._onDetachCallbacks.callCallbacks(this);
+      return this;
+    }
 
-      if (parent != null) {
-        parent.remove(this);
-      }
-      this._isAnchored = false;
-      this._parentElement = null;
+    /**
+     * Adds a callback to be called when th Component is detach()-ed.
+     *
+     * @param {ComponentCallback} callback The callback to be added.
+     * @return {Component} The calling Component.
+     */
+    public onDetach(callback: ComponentCallback) {
```

Comment:
The reason behind my choice of `onAnchor()`: For `onAnchor`, calling it immediately was a sensible choice. Also, it avoided a lot of hustle of checking the `_isAnchored` flag and made the code prettier.

Here the semantics are a bit different. A component that has not been attached yet is basically detached, but that does not mean that it was detached. To draw a parallel, a component that is anchored means that the component has gone through the anchor process, hence calling the callback is sensible. The difference is the initial state.